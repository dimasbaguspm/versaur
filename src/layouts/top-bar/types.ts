import type { HTMLAttributes } from 'react'

/**
 * TopBarRoot - main container for TopBar layout
 * Provides horizontal layout for desktop navigation/header
 * Compound pattern: Leading, Trailing, NavItem
 */
export type TopBarProps = HTMLAttributes<HTMLDivElement>

/**
 * TopBarLeading - left section (logo, brand, nav)
 */
export interface TopBarLeadingProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Custom class for leading section
   */
  className?: string
}

/**
 * TopBarTrailing - right section (actions, profile, etc)
 */
export interface TopBarTrailingProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Custom class for trailing section
   */
  className?: string
}

/**
 * TopBarNavItem - navigation item (link, button, etc)
 */
export interface TopBarNavItemProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * Whether the nav item is active
   */
  active?: boolean
  /**
   * Custom class for nav item
   */
  className?: string
}
