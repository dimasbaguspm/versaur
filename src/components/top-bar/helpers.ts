import { cva } from 'class-variance-authority'

/**
 * TopBarRoot styles
 */
export const topBarRootStyles = cva(
  'w-full flex items-center px-6 py-2 mb-6 border-b border-border shadow-xs bg-white',
  {
    variants: {},
    defaultVariants: {},
  }
)

/**
 * TopBarLeading styles
 */
export const topBarLeadingStyles = cva('flex items-center gap-3 min-w-0 flex-1')

/**
 * TopBarTrailing styles
 */
export const topBarTrailingStyles = cva('flex items-center gap-2 ml-auto')

/**
 * TopBarNavItem styles
 */
export const topBarNavItemStyles = cva(
  'px-2.5 py-1.5 rounded-md text-xs font-sm cursor-pointer transition-colors',
  {
    variants: {
      active: {
        true: 'bg-coral/10 text-coral',
        false: 'bg-neutral hover:bg-coral/10 text-ghost',
      },
    },
    defaultVariants: {
      active: false,
    },
  }
)

/**
 * TopBarNav styles (container for navigation items)
 */
export const topBarNavStyles = cva('flex items-center gap-2 ml-4')

/**
 * TopBarActions styles (container for action items)
 */
export const topBarActionsStyles = cva('flex items-center gap-2 mr-4')
