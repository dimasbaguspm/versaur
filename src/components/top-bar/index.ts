export { TopBar } from './top-bar'
export * from './types'
