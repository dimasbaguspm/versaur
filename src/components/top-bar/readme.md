# TopBar

A desktop-only compound layout header/navigation bar for Versaur UI. Follows the design system and code style conventions. Inspired by GitHub's TopBar, but adapted for Versaur's color palette, spacing, and accessibility.

## Usage

```tsx
<TopBar variant="primary">
  <TopBar.Leading>
    {/* Logo, brand, navigation */}
    <TopBar.Nav>
      <TopBar.NavItem active>Home</TopBar.NavItem>
      <TopBar.NavItem>Docs</TopBar.NavItem>
      <TopBar.NavItem>Pricing</TopBar.NavItem>
    </TopBar.Nav>
  </TopBar.Leading>
  <TopBar.Trailing>
    <TopBar.Actions>
      {/* Actions, profile, etc */}
      <ButtonIcon aria-label="Notifications" />
      <ButtonIcon aria-label="Settings" />
    </TopBar.Actions>
    <Avatar size="sm">DM</Avatar>
  </TopBar.Trailing>
</TopBar>
```

### Example: Navigation
```tsx
<TopBar>
  <TopBar.Leading>
    <TopBar.Nav>
      <TopBar.NavItem active>Home</TopBar.NavItem>
      <TopBar.NavItem>Docs</TopBar.NavItem>
      <TopBar.NavItem>Pricing</TopBar.NavItem>
    </TopBar.Nav>
  </TopBar.Leading>
  <TopBar.Trailing>
    <TopBar.Actions>
      <TopBar.NavItem>Sign In</TopBar.NavItem>
    </TopBar.Actions>
  </TopBar.Trailing>
</TopBar>
```

## API

- `TopBar`: Main container
- `TopBar.Leading`: Left section (logo, brand, navigation)
- `TopBar.Trailing`: Right section (actions, profile, etc)
- `TopBar.Nav`: Container for navigation items
- `TopBar.NavItem`: Individual navigation item
- `TopBar.Actions`: Container for action items

## Accessibility

- Uses semantic elements (`header`, `nav`)
- `aria-current="page"` for active navigation item
- Follows WCAG 2.1 AA color and contrast standards
