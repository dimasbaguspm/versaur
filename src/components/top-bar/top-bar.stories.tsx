/**
 * TopBar stories for Storybook
 * Group: Layout
 * Demonstrates TopBar compound usage, variants, and navigation items
 * Icons from lucide-react (dev only)
 */
import { BellIcon, BetweenHorizonalStartIcon, ZapIcon } from 'lucide-react'
import { Avatar } from '../avatar'
import { ButtonIcon } from '../button-icon'
import { TopBar } from './index'

export default {
  title: 'Layout/TopBar',
  component: TopBar,
  parameters: {
    layout: 'fullscreen',
  },
}

export const Default = () => (
  <TopBar>
    <TopBar.Leading>
      <Avatar size='sm' shape='rounded' variant='primary'>
        <ZapIcon />
      </Avatar>
      <TopBar.Nav>
        <TopBar.NavItem active>Home</TopBar.NavItem>
        <TopBar.NavItem>Docs</TopBar.NavItem>
        <TopBar.NavItem>Pricing</TopBar.NavItem>
      </TopBar.Nav>
    </TopBar.Leading>
    <TopBar.Trailing>
      <TopBar.Actions>
        <ButtonIcon
          as={BellIcon}
          size='sm'
          aria-label='Notifications'
          variant='outline'
        />
        <ButtonIcon
          as={BetweenHorizonalStartIcon}
          size='sm'
          aria-label='Settings'
          variant='outline'
        />
      </TopBar.Actions>
      <Avatar size='sm'>DM</Avatar>
    </TopBar.Trailing>
  </TopBar>
)
