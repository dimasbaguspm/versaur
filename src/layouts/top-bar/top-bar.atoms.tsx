import { forwardRef } from 'react'
import {
  topBarLeadingStyles,
  topBarTrailingStyles,
  topBarNavItemStyles,
  topBarNavStyles,
  topBarActionsStyles,
} from './helpers'
import type {
  TopBarLeadingProps,
  TopBarTrailingProps,
  TopBarNavItemProps,
} from './types'

export const TopBarNav = forwardRef<HTMLDivElement, TopBarLeadingProps>(
  function TopBarNav({ children, className, ...props }, ref) {
    return (
      <nav ref={ref} className={topBarNavStyles({ className })} {...props}>
        {children}
      </nav>
    )
  }
)

export const TopBarActions = forwardRef<HTMLDivElement, TopBarTrailingProps>(
  function TopBarActions({ children, className, ...props }, ref) {
    return (
      <div ref={ref} className={topBarActionsStyles({ className })} {...props}>
        {children}
      </div>
    )
  }
)

export const TopBarLeading = forwardRef<HTMLDivElement, TopBarLeadingProps>(
  function TopBarLeading({ children, className, ...props }, ref) {
    return (
      <div ref={ref} className={topBarLeadingStyles({ className })} {...props}>
        {children}
      </div>
    )
  }
)

export const TopBarTrailing = forwardRef<HTMLDivElement, TopBarTrailingProps>(
  function TopBarTrailing({ children, className, ...props }, ref) {
    return (
      <div ref={ref} className={topBarTrailingStyles({ className })} {...props}>
        {children}
      </div>
    )
  }
)

export const TopBarNavItem = forwardRef<HTMLDivElement, TopBarNavItemProps>(
  function TopBarNavItem({ children, className, active, ...props }, ref) {
    return (
      <div
        ref={ref}
        className={topBarNavItemStyles({ active, className })}
        aria-current={active ? 'page' : undefined}
        {...props}
      >
        {children}
      </div>
    )
  }
)
