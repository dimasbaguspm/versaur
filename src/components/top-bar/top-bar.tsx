import { forwardRef } from 'react'
import type { TopBarProps } from './types'
import { topBarRootStyles } from './helpers'

import {
  TopBarLeading,
  TopBarTrailing,
  TopBarNavItem,
  TopBarNav,
  TopBarActions,
} from './top-bar.atoms'

const TopBarRoot = forwardRef<HTMLDivElement, TopBarProps>(
  ({ children, className, ...props }, ref) => {
    return (
      <header ref={ref} className={topBarRootStyles({ className })} {...props}>
        {children}
      </header>
    )
  }
)

export const TopBar = Object.assign(TopBarRoot, {
  Leading: TopBarLeading,
  Trailing: TopBarTrailing,
  NavItem: TopBarNavItem,
  Nav: TopBarNav,
  Actions: TopBarActions,
})
