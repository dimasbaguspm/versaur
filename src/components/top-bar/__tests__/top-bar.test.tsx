import { render, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../top-bar.stories'

describe('TopBar', () => {
  const { Default } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders navigation items', () => {
    render(<Default />)
    expect(screen.getByText('Home')).toBeInTheDocument()
    expect(screen.getByText('Docs')).toBeInTheDocument()
    expect(screen.getByText('Pricing')).toBeInTheDocument()
  })

  it('renders actions and avatar', () => {
    render(<Default />)
    expect(screen.getByLabelText('Notifications')).toBeInTheDocument()
    expect(screen.getByLabelText('Settings')).toBeInTheDocument()
    expect(screen.getByText('DM')).toBeInTheDocument()
  })

  it('applies active state to nav item', () => {
    render(<Default />)
    const activeNav = screen.getByText('Home')
    expect(activeNav).toHaveAttribute('aria-current', 'page')
  })
})
