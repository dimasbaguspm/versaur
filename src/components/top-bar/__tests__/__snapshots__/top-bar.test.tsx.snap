// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopBar > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <header
    class="w-full flex items-center px-6 py-4 mb-6 border-b border-border shadow-xs bg-white"
  >
    <div
      class="flex items-center gap-3 min-w-0 flex-1"
    >
      <div
        class="inline-flex items-center justify-center font-medium overflow-hidden transition-all duration-200 select-none relative bg-coral-soft text-coral h-8 w-8 text-xs rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-zap"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
          />
        </svg>
      </div>
      <nav
        class="flex items-center gap-2 ml-4"
      >
        <div
          aria-current="page"
          class="px-2.5 py-1.5 rounded-md text-sm cursor-pointer transition-colors bg-coral/10 text-coral"
        >
          Home
        </div>
        <div
          class="px-2.5 py-1.5 rounded-md text-sm cursor-pointer transition-colors hover:text-coral text-ghost"
        >
          Docs
        </div>
        <div
          class="px-2.5 py-1.5 rounded-md text-sm cursor-pointer transition-colors hover:text-coral text-ghost"
        >
          Pricing
        </div>
      </nav>
    </div>
    <div
      class="flex items-center gap-2 ml-auto"
    >
      <div
        class="flex items-center gap-2 mr-4"
      >
        <button
          aria-disabled="false"
          aria-label="Notifications"
          class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-white text-foreground hover:bg-slate-soft focus-visible:ring-slate-soft focus-visible:ring-offset-white h-8 w-8 rounded-md"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-bell inline-flex items-center justify-center w-4 h-4 text-ghost"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.268 21a2 2 0 0 0 3.464 0"
            />
            <path
              d="M3.262 15.326A1 1 0 0 0 4 17h16a1 1 0 0 0 .74-1.673C19.41 13.956 18 12.499 18 8A6 6 0 0 0 6 8c0 4.499-1.411 5.956-2.738 7.326"
            />
          </svg>
        </button>
        <button
          aria-disabled="false"
          aria-label="Settings"
          class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-white text-foreground hover:bg-slate-soft focus-visible:ring-slate-soft focus-visible:ring-offset-white h-8 w-8 rounded-md"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-between-horizontal-start inline-flex items-center justify-center w-4 h-4 text-ghost"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="7"
              rx="1"
              width="13"
              x="8"
              y="3"
            />
            <path
              d="m2 9 3 3-3 3"
            />
            <rect
              height="7"
              rx="1"
              width="13"
              x="8"
              y="14"
            />
          </svg>
        </button>
      </div>
      <div
        class="inline-flex items-center justify-center font-medium overflow-hidden transition-all duration-200 select-none relative bg-coral-soft text-coral h-8 w-8 text-xs rounded-full"
      >
        DM
      </div>
    </div>
  </header>
</DocumentFragment>
`;
